### PR TITLE
177492288 report by asn

### PR DESCRIPTION
**Pull request by maciejkocylapc**
#### What's this PR do?
Add sort by asn option to data center page

#### How should this be manually tested?
There should be separate tab in data centers page for sorting by ASN

Note: 
In order to see any results you should run append_ip_geo_data.rb script first (remove LIMIT 10 in script if you can't see any results)

#### What are the relevant tickets?
https://www.pivotaltracker.com/story/show/177492288

#### Task completed checklist:
- [Y] Is there appropriate test coverage?
- [Y] Did I take into consideration possible security vulnerabilities?
- [Y] Are the controllers secure?
- [Y] Was every important matter documented?

#### Things to watch out for
- [N] Does this PR have any migrations?
- [N] Does this PR add new dependencies?

### CODE REVIEW

**Review by radek-polcode — CHANGES_REQUESTED** [review]
Please add new lines at the end of the files where needed.
Do we want to add epoch_wall_clocks in schema.rb?

File: app/views/data_centers/index.html.erb [hunk 1]
```diff
@@ -1,35 +1,25 @@
 <h1 class="page-header"><%= params[:network].capitalize %> Data Centers</h1>
-
+<div class=" mb-4 col-sm-4" role="toolbar" aria-label="filter"> 
+  <%= link_to "Data Centers", \
+              data_centers_url(network: params[:network], sort_by: :data_center), \
+              class: "btn #{[nil, 'data_center'].include?(@sort_by) ? 'btn-primary' : 'btn-outline-primary'}", \
```
Comment (kbiala) [inline app/views/data_centers/index.html.erb:5]: Please do not use ```btn-outline-primary``` class, it's not fully styled in app (has some bootstrap left-overs that we don't want). Use ```btn-secondary``` instead.

File: app/views/data_centers/index.html.erb [hunk 2]
```diff
@@ -1,35 +1,25 @@
...
+              type: "button" \
+  %>
+  <%= link_to "ASN",\
+              data_centers_url(network: params[:network], sort_by: :asn), \
+              class: "btn #{@sort_by == 'asn' ? 'btn-primary' : 'btn-outline-primary'}", \
```
Comment (kbiala) [inline app/views/data_centers/index.html.erb:10]: Same as above.

File: app/views/data_centers/index.html.erb (hunk 1 above)
Comment (kbiala) [reply app/views/data_centers/index.html.erb:5]: Also, please add ```mr-1``` class to all buttons

**Review by kbiala — CHANGES_REQUESTED** [review]

File: app/views/asns/show.html.erb [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+              <strong> ASN: </strong>
+              <%= params[:asn] %>
+            </th>
+          </tr>
+        </thead>
+        <tbody>
+        <tr>
+          <td class="category-column"><strong>Data Centers</strong></td>
+          <td>
+            <% @data_centers.uniq.each do |dc| %>
```
Comment (radek-polcode) [inline app/views/asns/show.html.erb:19]: `uniq` can be moved to controller I think.

File: app/views/validators/_as_row.html.erb [hunk 4]
```diff
@@ -0,0 +1,43 @@
+<% # EXAMPLE: render '/validators/as_row', score: ValidatorScoreV1.first %> 
+
+<% this_validator = score.validator %>
+<tr>
+  <td class="d-none d-lg-table-cell">
+    <span class="avatar">
+      <% if this_validator && this_validator.avatar_url %>
```
Comment (radek-polcode) [inline app/views/validators/_as_row.html.erb:7]: This can be moved to helper.

File: app/views/validators/_as_row.html.erb [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+                class: 'word-break-md' %>
+    <br />
+    <%= render 'validators/scores', validator: this_validator %>
+  </td>
+  <td>
+    <div class="mb-2">
+      <div class="d-lg-none">IP Address: </div>
+      <%= score.ip_address %>
+    </div>
+    <% if Ip.where(address: score.ip_address).first %>
```
Comment (radek-polcode) [inline app/views/validators/_as_row.html.erb:38]: Is it possible to pass this as a variable to this partial?

File: app/views/validators/_as_row.html.erb [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+      <div class="d-lg-none">Data Center Gateway: </div>
+      <%= Ip.where(address: score.ip_address).first&.data_center_host %>
```
Comment (radek-polcode) [inline app/views/validators/_as_row.html.erb:40]: Repetition

File: app/views/validators/_as_row.html.erb [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+    <% end %>
+  </td>
+  <td>
+    <div class="mb-2">
+      <div class="d-lg-none">Name & Stake:</div>
+      <%= link_to validator_display_name(this_validator) + ":",
+                  validator_url(network: params[:network], account: this_validator.account),
+                  class: 'font-weight-bold' %>
+      <%= number_with_delimiter lamports_to_sol(score.active_stake).round(2) %>
+      (<%= number_to_percentage((score.active_stake / @total_stake.to_f)*100.0, precision: 2) %>)
```
Comment (radek-polcode) [inline app/views/validators/_as_row.html.erb:24]: `(score.active_stake / @total_stake.to_f)*100.0` could be moved to the helper or maybe to a model if possible.

**Review by radek-polcode — CHANGES_REQUESTED** [review]
add missing new lines as well

File: app/views/data_centers/_asn_list.html.erb [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+    <td>
+      <% asn_data[:data_centers].each do |dc| %>
+        <%= link_to dc, data_center_url(network: params[:network], key: dc.gsub('/', '-slash-')) %>
+        <br \>
+      <% end %>
+    </td>
+    <td><%= asn_data[:count] %></td>
+    <td>
+      <%= number_with_delimiter lamports_to_sol(asn_data[:active_stake]).round(0) %>
+      (<%= number_to_percentage((asn_data[:active_stake] / @results[:total_stake].to_f) * 100.0, precision: 2) %>)
```
Comment (radek-polcode) [inline app/views/data_centers/_asn_list.html.erb:13]: This calculation `(asn_data[:active_stake] / @results[:total_stake].to_f) * 100.0` also could be in a helper.

File: app/views/data_centers/_data_center_list.html.erb [hunk 9]
```diff
@@ -0,0 +1,11 @@
+<% @results[:results].each do |dc_key, dc_data| %>
+  <tr>
+    <td><%= link_to dc_key, data_center_url(network: params[:network], key: dc_key.gsub('/', '-slash-')) %></td>
+    <td><%= dc_data[:aso] %></td>
+    <td><%= dc_data[:count] %></td>
+    <td>
+      <%= number_with_delimiter lamports_to_sol(dc_data[:active_stake]).round(0) %>
+      (<%= number_to_percentage((dc_data[:active_stake] / @results[:total_stake].to_f) * 100.0, precision: 2) %>)
```
Comment (radek-polcode) [inline app/views/data_centers/_data_center_list.html.erb:8]: `(dc_data[:active_stake] / @results[:total_stake].to_f) * 100.0` could be a helper method

File: app/controllers/asns_controller.rb [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+
+    render file: "#{Rails.root}/public/404.html", status: 404 if @data_centers.empty?
+
+    @scores = ValidatorScoreV1.where(network: asn_params[:network])
+                              .where('active_stake > 0')
+                              .where(data_center_key: @data_centers)
+
+    @asn_stake = @scores.sum(:active_stake)
+
+    @total_stake = ValidatorScoreV1.where(network: asn_params[:network])
```
[the remark this answers is not in the document]
Comment (maciejkocylapc) [reply app/controllers/asns_controller.rb:14]: we do not need validator here at all

File: app/controllers/asns_controller.rb [hunk 11]
```diff
@@ -1,18 +1,17 @@
...
     @data_centers = Ip.where(traits_autonomous_system_number: asn_params[:asn])
-                     .pluck(:data_center_key)
+                     .pluck(:data_center_key).uniq
 
     render file: "#{Rails.root}/public/404.html", status: 404 if @data_centers.empty?
 
-    @scores = ValidatorScoreV1.where(network: asn_params[:network])
-                              .where('active_stake > 0')
+    @scores = ValidatorScoreV1.includes(:validator).with_active_stake(asn_params[:network])
                               .where(data_center_key: @data_centers)
```
Comment (radek-polcode) [inline app/controllers/asns_controller.rb:9]: Consider make it a scope.

File: app/views/validators/_as_row.html.erb [hunk 12]
```diff
@@ -4,7 +4,7 @@
 <tr>
   <td class="d-none d-lg-table-cell">
     <span class="avatar">
-      <% if this_validator && this_validator.avatar_url %>
+      <% if avatar_present?(this_validator) %>
```
Comment (radek-polcode) [inline app/views/validators/_as_row.html.erb:7]: Make all condition a helper method.

**Review by radek-polcode — CHANGES_REQUESTED** [review]

File: test/factories/ips.rb [hunk 13]
```diff
@@ -1,5 +1,35 @@
...
+    registered_country_name { 'China' }
+    traits_anonymous { true }
+    traits_hosting_provider { true }
+    traits_user_type { 'hosting' }
+    traits_autonomous_system_number { 00000 }
+    traits_isp { 'Organization' }
+    traits_organization { 'Organization' }
+    location_time_zone { 'Asia/Shanghai' }
+
+    factory :ip_berlin do
```
Comment (radek-polcode) [inline test/factories/ips.rb:21]: I think you shoul use trait here and below https://github.com/thoughtbot/factory_bot/blob/master/GETTING_STARTED.md#traits

**Review by radek-polcode — APPROVED** [review]

**Pull request merged**